receiver/signalfxreceiver/signalfxv2_to_metricdata.go

```diff
@@ -0,0 +1,238 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package signalfxreceiver
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	sfxpb "github.com/signalfx/com_signalfx_metrics_protobuf"
+	"go.uber.org/zap"
+)
+
+var (
+	errSFxNilDatum                   = errors.New("nil datum value for data-point")
+	errSFxMetricTypeEnumNotSupported = errors.New("data-point of MetricType_ENUM type is not supported")
+
+	errSFxUnexpectedInt64DatumType   = errors.New("datum value type of int64 is unexpected")
+	errSFxUnexpectedFloat64DatumType = errors.New("datum value type of float64 is unexpected")
+	errSFxUnexpectedStringDatumType  = errors.New("datum value type of string is unexpected")
+	errSFxStringDatumNotNumber       = errors.New("datum string cannot be parsed to a number")
+	errSFxNoDatumValue               = errors.New("no datum value present for data-point")
+)
+
+// SignalFxV2ToMetricsData converts SignalFx proto data points to
+// consumerdata.MetricsData. Returning the converted data and the number of
+// dropped time series.
+func SignalFxV2ToMetricsData(
+	logger *zap.Logger,
+	sfxDataPoints []*sfxpb.DataPoint,
+) (*consumerdata.MetricsData, int) {
+
+	// TODO: not optimized at all, basically regenerating everything for each
+	// 	data point.
+	numDroppedTimeSeries := 0
+	md := &consumerdata.MetricsData{}
+	metrics := make([]*metricspb.Metric, 0, len(sfxDataPoints))
+	for _, sfxDataPoint := range sfxDataPoints {
+		if sfxDataPoint == nil {
+			// TODO: Log or metric for this odd ball?
+			continue
+		}
+
+		// First check if the type is convertible and the data point is consistent.
+		metricType, err := convertType(sfxDataPoint)
+		if err != nil {
+			numDroppedTimeSeries++
+			logger.Debug("SignalFx data-point type conversion error",
+				zap.Error(err),
+				zap.String("metric", sfxDataPoint.GetMetric()))
+			continue
+		}
+		point, err := buildPoint(sfxDataPoint, metricType)
+		if err != nil {
+			numDroppedTimeSeries++
+			logger.Debug("SignalFx data-point datum conversion error",
+				zap.Error(err),
+				zap.String("metric", sfxDataPoint.GetMetric()))
+			continue
+		}
+
+		labelKeys, labelValues := buildLabelKeysAndValues(sfxDataPoint.Dimensions)
+		descriptor := buildDescriptor(sfxDataPoint, labelKeys, metricType)
+		ts := &metricspb.TimeSeries{
+			// TODO: StartTimestamp can be set if each cumulative time series are
+			//  	tracked but right now it is not clear if it brings benefits.
+			//		Perhaps as an option so cost is "pay for play".
+			LabelValues: labelValues,
+			Points:      []*metricspb.Point{point},
+		}
+		metric := &metricspb.Metric{
```

Comment:
How about `sfxDataPoint.Properties` field? Do we ignore this? I think this corresponds to `Resource` attributes (I may be wrong).

Reply:
The type that I'm using is from proto and it doesn't have `Properties`:
```golang
type DataPoint struct {
	Source           *string      `protobuf:"bytes,1,opt,name=source" json:"source,omitempty"`
	Metric           *string      `protobuf:"bytes,2,opt,name=metric" json:"metric,omitempty"`
	Timestamp        *int64       `protobuf:"varint,3,opt,name=timestamp" json:"timestamp,omitempty"`
	Value            *Datum       `protobuf:"bytes,4,opt,name=value" json:"value,omitempty"`
	MetricType       *MetricType  `protobuf:"varint,5,opt,name=metricType,enum=com.signalfx.metrics.protobuf.MetricType" json:"metricType,omitempty"`
	Dimensions       []*Dimension `protobuf:"bytes,6,rep,name=dimensions" json:"dimensions,omitempty"`
	XXX_unrecognized []byte       `json:"-"`
}
```
I assume that `Properties` that you mention comes from the client lib, let me see what it is translated to.

Reply:
OK, ignore this.

Looks like it was removed from Protobuf definition https://github.com/signalfx/com_signalfx_metrics_protobuf/commit/5e41ba1725d48aba7aa292d3ee4d0c0e6ddd6937

I just had an old version of SmartAgent source code which was refering to the version of the Protobuf which had this field.

However, this makes it unclear to me how are resource attributes reported? `Dimensions` are supposedly the identity of the metric, i.e. the label keys. Where do resource attributes like hostname go? Also in `Dimensions`?

Reply:
I found `Properties` on the event type:
```golang
type Event struct {
	EventType        *string        `protobuf:"bytes,1,req,name=eventType" json:"eventType,omitempty"`
	Dimensions       []*Dimension   `protobuf:"bytes,2,rep,name=dimensions" json:"dimensions,omitempty"`
	Properties       []*Property    `protobuf:"bytes,3,rep,name=properties" json:"properties,omitempty"`
	Category         *EventCategory `protobuf:"varint,4,opt,name=category,enum=com.signalfx.metrics.protobuf.EventCategory" json:"category,omitempty"`
	Timestamp        *int64         `protobuf:"varint,5,opt,name=timestamp" json:"timestamp,omitempty"`
	XXX_unrecognized []byte         `json:"-"`
}
```

Reply:
Perhaps check with SFx folks.